src/runtime_src/core/edge/drm/zocl/edge/zocl_hwctx.c

```diff
@@ -0,0 +1,460 @@
+    else
+        ret = -EINVAL;
+
+    mutex_unlock(&client->lock);
+    return ret;
+}
+
+int zocl_hw_ctx_execbuf(struct drm_zocl_dev *zdev, struct drm_zocl_hw_ctx_execbuf *drm_hw_ctx_execbuf, struct drm_file *filp)
+{
+    struct drm_gem_object *gem_obj = NULL;
+    struct drm_device *dev = zdev->ddev;
+    struct kds_client *client = filp->driver_priv;
+    struct drm_zocl_bo *zocl_bo = NULL;
+    struct ert_packet *ecmd = NULL;
+    struct kds_command *xcmd = NULL;
+    int ret = 0;
+
+    if (zdev->kds.bad_state) {
+        DRM_ERROR("%s: KDS is in bad state", __func__);
+        return -EDEADLK;
+    }
+    gem_obj = zocl_gem_object_lookup(dev, filp, drm_hw_ctx_execbuf->exec_bo_handle);
+    if (!gem_obj) {
+        DRM_ERROR("%s: Look up GEM BO %d failed", __func__, drm_hw_ctx_execbuf->exec_bo_handle);
+        return -EINVAL;
+    }
+
+    zocl_bo = to_zocl_bo(gem_obj);
+    if (!zocl_bo_execbuf(zocl_bo)) {
+        DRM_ERROR("%s: Command Buffer is not exec buf", __func__);
+        return -EINVAL;
+    }
+
+    ecmd = (struct ert_packet *)zocl_bo->cma_base.vaddr;
+    ecmd->state = ERT_CMD_STATE_NEW;
+
+    xcmd = kds_alloc_command(client, ecmd->count * sizeof(u32));
+    if (!xcmd) {
+        DRM_ERROR("%s: Failed to alloc xcmd", __func__);
+        ret = -ENOMEM;
+        goto out;
+    }
+
+    xcmd->cb.free = kds_free_command;
+    xcmd->cb.notify_host = notify_execbuf;
+    xcmd->execbuf = (u32 *)ecmd;
+    xcmd->gem_obj = gem_obj;
+    xcmd->exec_bo_handle = drm_hw_ctx_execbuf->exec_bo_handle;
+    xcmd->hw_ctx_id = drm_hw_ctx_execbuf->hw_ctx_id;
+
+    switch (ecmd->opcode) {
+        case ERT_CONFIGURE:
+            xcmd->status = KDS_COMPLETED;
+            xcmd->cb.notify_host(xcmd, xcmd->status);
+            goto out1;
+        case ERT_START_CU:
+            start_krnl_ecmd2xcmd(to_start_krnl_pkg(ecmd), xcmd);
+            break;
+        case ERT_EXEC_WRITE:
+            DRM_WARN_ONCE("ERT_EXEC_WRITE is obsoleted, use ERT_START_KEY_VAL");
+#if KERNEL_VERSION(5, 4, 0) > LINUX_VERSION_CODE
+		    __attribute__ ((fallthrough));
+#else
+		    __attribute__ ((__fallthrough__));
+#endif
+        case ERT_START_KEY_VAL:
+            start_krnl_kv_ecmd2xcmd(to_start_krnl_pkg(ecmd), xcmd);
+            break;
+        case ERT_START_FA:
+            start_fa_ecmd2xcmd(to_start_krnl_pkg(ecmd), xcmd);
+            break;
+        case ERT_START_COPYBO:
+            ret = copybo_ecmd2xcmd(zdev, filp, to_copybo_pkg(ecmd), xcmd);
+            if (ret)
+                goto out1;
+            goto out;
+        case ERT_ABORT:
+            abort_ecmd2xcmd(to_abort_pkg(ecmd), xcmd);
+            break;
+        default:
+            DRM_ERROR("%s: Unsupport command", __func__);
+            ret = -EINVAL;
+            goto out1;
+    }
+
+    if (check_for_open_hw_cu_ctx(zdev, client, xcmd) < 0) {
+        DRM_ERROR("The client has no opening context\n");
+        return -EINVAL;
+    }
+
+    ret = kds_add_command(&zdev->kds, xcmd);
```

Comment:
```suggestion
    return kds_add_command(&zdev->kds, xcmd);
```

Reply:
Hi Saif, made the necessary changes here with the latest push. Thanks